`four-forward-slashes` thinks module-level docs are items that can be documented

Using the following flags

`--force-warn clippy::four-forward-slashes`

this code:

```rust
pub fn main() {
    //// I am not a doc comment!
    //! another comment
}

```

caused the following diagnostics:

```
    Checking _a v0.1.0 (/tmp/icemaker_global_tempdir.ZJcEP8tl8ptu/icemaker_clippyfix_tempdir.U1dWplOlXAU9/_a)
warning: this item has comments with 4 forward slashes (`////`). These look like doc comments, but they aren't
 --> src/main.rs:1:1
  |
1 | / pub fn main() {
2 | |     //// I am not a doc comment!
3 | |     //! another comment
  | |_______________________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#four_forward_slashes
  = note: requested on the command line with `--force-warn clippy::four-forward-slashes`
help: make this a doc comment by removing one `/`
  |
2 -     //// I am not a doc comment!
3 -     //! another comment
2 + /// I am not a doc comment!
  |

warning: `_a` (bin "_a") generated 1 warning (run `cargo clippy --fix --bin "_a" -p _a` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s

```

However after applying these diagnostics, the resulting code:

```rust
pub fn main() {
/// I am not a doc comment!
    // another comment
}

```
no longer compiled:

```
    Checking _a v0.1.0 (/tmp/icemaker_global_tempdir.ZJcEP8tl8ptu/icemaker_clippyfix_tempdir.U1dWplOlXAU9/_a)
error[E0585]: found a documentation comment that doesn't document anything
 --> src/main.rs:2:1
  |
2 | /// I am not a doc comment!
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: doc comments must come before what they document, if a comment was intended use `//`

For more information about this error, try `rustc --explain E0585`.
error: could not compile `_a` (bin "_a") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_a` (bin "_a" test) due to 1 previous error

```

Version:
```
rustc 1.93.0-nightly (9b82a4fff 2025-12-01)
binary: rustc
commit-hash: 9b82a4fffe2b215f488a6dfdc0b508235f37c85c
commit-date: 2025-12-01
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5

```

--- a/clippy_lints/src/four_forward_slashes.rs
+++ b/clippy_lints/src/four_forward_slashes.rs
@@ -1,8 +1,10 @@
 use clippy_utils::diagnostics::span_lint_and_then;
 use clippy_utils::source::SpanExt as _;
 use itertools::Itertools as _;
+use rustc_ast::AttrStyle;
 use rustc_errors::Applicability;
-use rustc_hir::Item;
+use rustc_hir::attrs::AttributeKind;
+use rustc_hir::{Attribute, Item};
 use rustc_lint::{LateContext, LateLintPass, LintContext as _};
 use rustc_session::declare_lint_pass;
 use rustc_span::Span;
@@ -48,7 +50,19 @@ impl<'tcx> LateLintPass<'tcx> for FourForwardSlashes {
             .tcx
             .hir_attrs(item.hir_id())
             .iter()
-            .filter(|i| i.is_doc_comment().is_some())
+            // Only fold in OUTER doc comments. An inner doc (`//!`) inside the item's
+            // body lowers onto the item with `AttrStyle::Inner`; folding its span drags
+            // `end_line` down into the body, so the upward `////` scan then flags a
+            // regular comment there (see #16168).
+            .filter(|i| {
+                matches!(
+                    i,
+                    Attribute::Parsed(AttributeKind::DocComment {
+                        style: AttrStyle::Outer,
+                        ..
+                    })
+                )
+            })
             .fold(item.span.shrink_to_lo(), |span, attr| span.to(attr.span()));
         let (Some(file), _, _, end_line, _) = sm.span_to_location_info(span) else {
             return;

--- a/tests/ui/four_forward_slashes.fixed
+++ b/tests/ui/four_forward_slashes.fixed
@@ -47,3 +47,10 @@ with_span! {
     //// don't lint me bozo
     fn f() {}
 }
+
+// Regression test for #16168: an inner doc comment (`//!`) inside the body must not
+// drag the `////` scan down into the body and flag a regular comment there.
+fn inner_doc() {
+    //// I am not a doc comment!
+    //! inner doc comment
+}

--- a/tests/ui/four_forward_slashes.rs
+++ b/tests/ui/four_forward_slashes.rs
@@ -47,3 +47,10 @@ with_span! {
     //// don't lint me bozo
     fn f() {}
 }
+
+// Regression test for #16168: an inner doc comment (`//!`) inside the body must not
+// drag the `////` scan down into the body and flag a regular comment there.
+fn inner_doc() {
+    //// I am not a doc comment!
+    //! inner doc comment
+}

